x-pack/plugins/reporting/server/core.ts

```diff
@@ -239,12 +241,32 @@ export class ReportingCore {
 
   /**
    * If xpack.reporting.roles.enabled === true, register Reporting as a feature
-   * that is controlled by user role names
+   * that is controlled by user role names. Also, for Serverless register a
+   * 'shell' Reporting Kibana feature.
    */
-  public registerFeature() {
+  public registerFeatures(isServerless: boolean) {
     const { features } = this.getPluginSetupDeps();
     const deprecatedRoles = this.getDeprecatedAllowedRoles();
 
+    // Register a 'shell' feature specifically for Serverless. If granted, it will automatically provide access to
```

Comment:
__note:__ I feel like this this "shell" definition belongs to the Reporting plugin, but let me know if you want me to move it somewhere else (e.g. `serverless` plugin).

Reply:
I feel like it belongs to the Reporting plugin, since the logic should be intertwined with Reporting-specific config: whether or not the deprecated roles setting is enabled makes a difference in how the feature gets registered.

Let's not have `reportingCore` used for this, instead let's move the `registerFeatures` logic to a separate utility function: `x-pack/plugins/reporting/server/features.ts` and call that function from here.

in plugin.ts:
```typescript
      // Feature registration relies on config, depending on whether deprecated roles are enabled, so it cannot be setup before here.
      registerFeatures({
        features: plugins.features,
        deprecatedRoles: reportingCore.getDeprecatedAllowedRoles(),
        isServerless: this.initContext.env.packageInfo.buildFlavor === 'serverless',
        logger: this.logger,
      });
```

in features.ts:
```typescript
import { DEFAULT_APP_CATEGORIES, type Logger } from '@kbn/core/server';
import { i18n } from '@kbn/i18n';
import type { PluginSetupContract as FeaturesPluginSetup } from '@kbn/features-plugin/server';

interface FeatureRegistrationOpts {
  features: FeaturesPluginSetup;
  deprecatedRoles: string[] | false;
  isServerless: boolean;
  logger: Logger;
}

/**
 * If xpack.reporting.roles.enabled === true, register Reporting as a feature
 * that is controlled by user role names. Also, for Serverless register a
 * 'shell' Reporting Kibana feature.
 */

export const registerFeatures = ({
  isServerless,
  features,
  deprecatedRoles,
  logger,
}: FeatureRegistrationOpts) => {
  // Register a 'shell' feature specifically for Serverless. If granted, it will automatically provide access to
  // reporting capabilities in other features, such as Discover, Dashboards, and Visualizations. On its own, this
  // feature doesn't grant any additional privileges.
  if (isServerless) {
    features.registerKibanaFeature({ ... });
  }
  ...
};
```

Reply:
Refactored. I kept the function in that async block as it was done before, even though I didn't manage to figure out why we would need that 🙂